virtool/uploads/data.py

```diff
@@ -0,0 +1,204 @@
+from logging import getLogger
+from typing import List, Optional, Union
+
+from sqlalchemy.ext.asyncio import AsyncSession
+from sqlalchemy import select, update
+from virtool_core.models.upload import UploadSearchResult, Upload
+from virtool_core.utils import rm
+
+import virtool.utils
+from virtool.data.errors import ResourceNotFoundError
+from virtool.data.piece import DataLayerPiece
+
+
+from virtool.mongo.core import DB
+from virtool.mongo.transforms import apply_transforms
+from virtool.uploads.db import finalize
+from virtool.uploads.models import Upload as SQLUpload
+from virtool.users.db import AttachUserTransform
+from virtool.utils import run_in_thread
+
+logger = getLogger(__name__)
+
+
+class UploadsData(DataLayerPiece):
+    def __init__(self, config, db, pg):
+        self._config = config
+        self._db: DB = db
+        self._pg = pg
+
+    async def find(
+        self, user: str = None, upload_type: str = None, ready: bool = None
+    ) -> UploadSearchResult:
+        """
+        Find and filter uploads.
+        """
+
+        filters = [SQLUpload.removed == False]
+        uploads = []
+
+        async with AsyncSession(self._pg) as session:
+            if user:
+                filters.append(SQLUpload.user == user)
+
+            if upload_type:
+                filters.append(SQLUpload.type == upload_type)
+
+            if ready is not None:
+                filters.append(SQLUpload.ready == ready)
+
+            results = await session.execute(select(SQLUpload).filter(*filters))
+
+        for result in results.unique().scalars().all():
+            uploads.append(result.to_dict())
+
+        uploads = await apply_transforms(uploads, [AttachUserTransform(self._db)])
+
+        return UploadSearchResult(documents=uploads)
+
+    async def create(
+        self,
+        name: str,
+        upload_type: str,
+        reserved: bool = False,
```

Comment:
Just pass in the model object as a positional argument called `data` like we do elsewhere.

It is a recipe for bugs to have default values on these function parameters in addition to the defaults and validation on the models.

The `user` parameter will have to stay as is because it is not part of the request model.